src/vm/vm_memory/memory.rs

```diff
@@ -111,73 +110,57 @@ impl Memory {
-                segment_data
-                    .into_iter()
-                    .enumerate()
-                    .map(move |(cell_offset, cell_data)| {
-                        (
-                            Relocatable::from((segment_index as isize, cell_offset)),
-                            cell_data,
-                        )
-                    })
-            })
-            .chain(prev_temp_data.into_iter().enumerate().flat_map(
-                |(segment_index, segment_data)| {
-                    let segment_index = -(segment_index as isize) - 1;
-                    segment_data
-                        .into_iter()
-                        .enumerate()
-                        .map(move |(cell_offset, cell_data)| {
-                            (Relocatable::from((segment_index, cell_offset)), cell_data)
-                        })
-                },
-            ));
-        for (addr, value) in data_iter {
-            // After the following check, addr.segment_index cannot be negative, therefore it is
-            // safe to cast to `usize`.
-            if addr.segment_index.is_negative() {
-                continue;
-            }
-
-            let value = match value {
-                Some(x) => x,
-                None => continue,
-            };
-
-            let new_addr: Relocatable = self
-                .relocate_value(&MaybeRelocatable::RelocatableValue(addr))
-                .into_owned()
-                .try_into()?;
-            let new_value = self.relocate_value(&value).into_owned();
-
-            if new_addr.segment_index as usize >= self.data.len() {
-                self.data
-                    .resize(new_addr.segment_index as usize + 1, Vec::new());
+        // Relocate temporary addresses in memory
+        for segment in self.data.iter_mut().chain(self.temp_data.iter_mut()) {
+            for value in segment.iter_mut() {
+                match value {
+                    Some(MaybeRelocatable::RelocatableValue(addr)) if addr.segment_index < 0 => {
+                        *value = Some(Memory::relocate_address(addr, &self.relocation_rules));
+                    }
+                    _ => {}
+                }
             }
-
-            let segment_data = &mut self.data[new_addr.segment_index as usize];
-            if new_addr.offset >= segment_data.len() {
-                segment_data.resize(new_addr.offset + 1, None);
+        }
+        // Move relocated temporary memory into the real memory
+        for index in (0..self.temp_data.len()).rev() {
+            if let Some(base_addr) = self.relocation_rules.get(&index) {
+                let data_segment = self.temp_data.remove(index);
+                // Insert the to-be relocated segment into the real memory
+                let mut addr = *base_addr;
+                if let Some(s) = self.data.get_mut(addr.segment_index as usize) {
+                    s.reserve_exact(data_segment.len())
+                }
```

Comment:
Won't `Memory::insert()` take care of reserving enough space? Specially since we're inserting reversed, the first insertion should be enough for that.

Reply:
I added the reserve in order to resize the memory segment to fit all of the elements in the temporary segment before inserting its elements one by one (as to minimize resizing)

Reply:
Yes, but the question is whether it'll be inserting by the end, essentially resizing only once anyway.

Reply:
We are inserting elements one by one incrementing de address with each insertion, so not really